source/fab/tools/versioning.py

```diff
@@ -0,0 +1,241 @@
+##############################################################################
+# (c) Crown copyright Met Office. All rights reserved.
+# For further details please refer to the file COPYRIGHT
+# which you should have received as part of this distribution
+##############################################################################
+
+"""This file contains the base class for versioning tools like git and
+subversion. It also contains derived classes Git, Subversion, and Fcm.
+"""
+
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+
+from fab.tools.categories import Categories
+from fab.tools.tool import Tool
+
+
+class Versioning(Tool):
+    '''This is the base class for versioning tools like git and svn.
+
+    :param name: the name of the tool.
+    :param exec_name: the name of the executable of this tool.
+    :param working_copy_command: which command is run to determine if
+        a directory is a working copy for this tool or not.
+    :param category: the category to which this tool belongs).
+    '''
+
+    def __init__(self, name: str,
+                 exec_name: str,
+                 working_copy_command: str,
+                 category: Categories):
+        super().__init__(name, exec_name, category)
+        self._working_copy_command = working_copy_command
+
+    def check_available(self) -> bool:
+        ''':returns: whether this tool is installed or not.'''
+        try:
+            self.run("help")
+        except RuntimeError:
+            return False
+        return True
+
+    def is_working_copy(self, path: Union[str, Path]) -> bool:
+        """:returns: whether the given path is a working copy or not. It
+            runs the command specific to the instance.
+
+        :param path: directory to be checked.
+        """
+        try:
+            self.run([self._working_copy_command], cwd=path,
+                     capture_output=False)
+        except RuntimeError:
+            return False
+        return True
```

Comment:
It is used in `steps/grab/{git,svn}` - and I am pretty sure I implemented it because these functions used it in the first place. I am happy to remove this (which means the grab steps will likely fail with an exception indicating that this is not  a working directory, which I am totally ok with :) )

Reply:
Lets have it out and reduce this size of this change by a few lines.

Reply:
Really? I am surprised :) OK, I have done it in a stand-alone commit so we can undo it in case that I misunderstood. So the svn and git steps do not check anymore if they are running on a correct repository anymore, instead this will now raise a RuntimeError (by svn/git).